### PR TITLE
proxmox: Add public key support: fix #19487

### LINKED ISSUE #19487: Add public key support to proxmox module
_From @DavidVentura on September 11, 2016 23:28_

Since proxmox 4.2 it's possible to create containers with a pre-defined public key to allow ssh access as root.
##### ISSUE TYPE
- Feature Idea
##### COMPONENT NAME

proxmox
##### ANSIBLE VERSION

```
ansible 2.1.0.0

```

I can currently create containers with

```
proxmox: vmid=$CT_NUMBER node='$NODE' api_user='david@pam' api_password='redacted' api_host='$API_HOST' password='$CT_PASS' hostname='$HOSTNAME' storage='redacted' ostemplate='especificos:vztmpl/debian-8.0-standard_8.4-1_amd64.tar.gz'
```

It'd be lovely to avoid having to manually attach to the new containers to add the pubkey before continuing with the process.

_Copied from original issue: ansible/ansible-modules-extras#2897_

### PR DESCRIPTION
**Pull request by DavidVentura**
##### ISSUE TYPE
Bugfix Pull Request

##### COMPONENT NAME
proxmox

##### ANSIBLE VERSION
2.3

##### SUMMARY

this PR fixes #19487 (adding pubkey support to proxmox module)

### CODE REVIEW

File: lib/ansible/modules/cloud/misc/proxmox.py [hunk 1]
```diff
@@ -175,6 +175,14 @@
      - Indicate desired state of the instance
     choices: ['present', 'started', 'absent', 'stopped', 'restarted']
     default: present
+  pubkey:
+    version_added: "2.3"
+    description:
+      - Public key to add to /root/.ssh/authorized_keys
```
Comment (gundalow) [inline lib/ansible/modules/cloud/misc/proxmox.py]: `Public key to add to .ssh/authorized_keys.`
Missing full stop
May not be run as root

File: lib/ansible/modules/cloud/misc/proxmox.py [hunk 2]
```diff
@@ -175,6 +175,14 @@
      - Indicate desired state of the instance
     choices: ['present', 'started', 'absent', 'stopped', 'restarted']
     default: present
+  pubkey:
+    version_added: "2.3"
+    description:
```
Comment (gundalow) [inline lib/ansible/modules/cloud/misc/proxmox.py]: Generally we put the `description:` first to make the file more readable.

**Review by gundalow — CHANGES_REQUESTED** [review]
needs_revision

**Comment by UnderGreen** [comment]
Will fail on ProxmoxVE < 4.2. You have to add backport compatibility.

**Comment by DavidVentura** [comment]
@UnderGreen It will only fail **only** if the options are specified and proxmox does not support it. You get an api response saying that the option is not supported. Is this behavior not ok?

File: lib/ansible/modules/cloud/misc/proxmox.py (hunk 1 above)
Comment (DavidVentura) [reply lib/ansible/modules/cloud/misc/proxmox.py]: Why can't it be run as root?

Comment (gundalow) [reply lib/ansible/modules/cloud/misc/proxmox.py]: Apologies, my comment wasn't clear.
I assume this module could be ran as root *or* another user. Therefore the `description:` should be something closer to
`Public key to add to users .ssh/authorized_keys.

**Comment by DavidVentura** [comment]
Proxmox adds the key to the root user upon creation

On December 19, 2016 1:52:17 PM GMT-03:00, John R Barker <notifications@github.com> wrote:
>gundalow commented on this pull request.
>
>
>
>> @@ -175,6 +175,14 @@
>      - Indicate desired state of the instance
>     choices: ['present', 'started', 'absent', 'stopped', 'restarted']
>     default: present
>+  pubkey:
>+    version_added: "2.3"
>+    description:
>+      - Public key to add to /root/.ssh/authorized_keys
>
>Apologies, my comment wasn't clear.
>I assume this module could be ran as root *or* another user. Therefore
>the `description:` should be something closer to
>`Public key to add to users .ssh/authorized_keys.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/ansible/ansible/pull/19515

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by UnderGreen** [comment]
@DavidVentura if Proxmox doesn't support something it will response with 500x :cry:

File: lib/ansible/modules/cloud/misc/proxmox.py (hunk 1 above)
Comment (DavidVentura) [reply lib/ansible/modules/cloud/misc/proxmox.py]: Proxmox adds the specified public key to the root user upon container creation.

File: lib/ansible/modules/cloud/misc/proxmox.py (hunk 2 above)
Comment (DavidVentura) [reply lib/ansible/modules/cloud/misc/proxmox.py]: Swapped in the last commit

**Comment by DavidVentura** [comment]
@UnderGreen Just pushed the version check

**Comment by DavidVentura** [comment]
I totally forgot about this. What do we do now? Do you want me to rebase this ?

**Comment by gundalow** [comment]
Please rebase

File: lib/ansible/modules/cloud/misc/proxmox.py [hunk 3]
```diff
@@ -175,6 +175,14 @@
      - Indicate desired state of the instance
     choices: ['present', 'started', 'absent', 'stopped', 'restarted']
     default: present
+  pubkey:
+    description:
+      - Public key to add to /root/.ssh/authorized_keys.
```
Comment (gundalow) [inline lib/ansible/modules/cloud/misc/proxmox.py]: If I'm reading the code correct, does this only work on proxmox 4.2+? If so may be worth stating that here.

File: lib/ansible/modules/cloud/misc/proxmox.py [hunk 4]
```diff
@@ -175,6 +175,14 @@
...
+    version_added: "2.3"
+    default: null
+    required: false
+    type: string
```
Comment (gundalow) [inline lib/ansible/modules/cloud/misc/proxmox.py]: No need to document `type:`, please remove.

**Comment by DavidVentura** [comment]
I seem to have broken my merge beyond human ability. I'll clean this up and open a new PR with a clean branch.

**Pull request closed without merging**